### ISSUE TITLE
API deprecation policy

### ISSUE THREAD

**Issue by s-urbaniak** [prometheus-operator/prometheus-operator#3439]
So far prometheus-operator has no formalized API deprecation policy or documented stability guarantees. Taking as the precedence the conversation in https://github.com/prometheus-operator/prometheus-operator/issues/3089 I suggest to formalize and document our API guarantees and make it part of our readme and potentially a central documentation piece.

A first suggestion is to follow the Kubernetes deprecation policy as documented in https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api which would also align to our declared `v1` CRD types.

@prometheus-operator/prometheus-operator-reviewers

**Comment by paulfantom** [comment]
SGTM after we release 1.x. Right now we are at 0.x.y and per sem-ver docs this means no stability guarantees for any interfaces.

**Comment by s-urbaniak** [comment]
I think it is not reasonable we change APIs today alrady, also see the comment in https://github.com/prometheus-operator/prometheus-operator/issues/3089#issuecomment-601168408. Prometheus-operator is in wide use today and even we we do have breaking API changes today we should at least have migration code.

**Comment by simonpasquier** [comment]
I'm not too familiar with CRD versioning but can we mix `v1` and `v1beta1` (or `v1alpha1`) for different resources? Case in point: #3451 adds a new AlertmanagerConfig CRD that might need some time to get right.

**Comment by s-urbaniak** [comment]
yes, they can definitely be mixed and that is how it is done in Kubernetes too. #3451 is actually a good precedence, we could declare it alpha first and then gradually promote to stable.

**Comment by s-urbaniak** [comment]
Having said that, Kubernetes has a "conversions" framework in place which is able to handle n-1 versioned object types.

**Comment by brancz** [comment]
> we could declare it alpha first and then gradually promote to stable.

This did happen, the APIs were alpha for a very long time. The APIs are v1, so we cannot break them in a backward compatible way, we can however, change the behavior, which is why the prometheus operator version has been non stable (to reflect that).

If we want to create a new iteration we will need to either do a v2alpha1 or an entirely new API (which probably makes sense for other reasons, like getting rid of the coreos API group).
